src/libraries/LiquidatorLib.sol

```diff
@@ -296,4 +355,172 @@ library LiquidatorLib {
+   * @param reserve1 The current reserve of token 1
+   * @param reserve0 The curernt reserve of token 0
+   * @param priceImpact1 The price impact to allow after swapping in the resulting amount
+   * @return amountIn1 The amount of token 1 to swap in to achieve the desired maximum price impact
+   */
+  function calculateRestrictedAmounts(
+    uint128 reserve1,
+    uint128 reserve0,
+    UFixed32x4 priceImpact1
+  ) internal pure returns (uint256 amountIn1, uint256 amountOut0) {
+    // If the price of token 1 is really small, do a different calculation to avoid rounding errors.
+    if (reserve1 > reserve0) {
+      return calculateRestrictedAmountsInverted(reserve1, reserve0, priceImpact1);
+    }
+
+    uint256 scalar = reserve0 - reserve1 >= type(uint88).max ? 1 : 1e8;
+
+    // "Price" isn't a very clear name. It's the amount of token 0 per token 1.
+    // Calculate p1
+    // p1 = r0 / r1
+    uint256 price1 = (uint256(reserve0) * scalar) / reserve1;
+
+    // Calculate p1'
+    // p1' = p1 - (p1 * priceImpact1)
+    uint256 price1_1 = price1 - FixedMathLib.mul(price1, priceImpact1);
+
+    // Calculate r0'
+    // p1' = r0' / r1'
+    // r1' = r0' / p1'
+
+    // r0' * r1' = r0 * r1
+    // r0' * (r0' / p1') = r0 * r1
+    // r0'^2 / p1' = r0 * r1
+    // r0'^2 = r0 * r1 * p1'
+    // r0' = sqrt(r0 * r1 * p1')
+    uint256 reserve0_1 = Math.sqrt(((uint256(reserve0) * reserve1 * price1_1) / scalar));
+
+    // Calculate r1'
+    // r0' * r1' = r0 * r1
+    // r1' = (r0 * r1) / r0'
+    uint256 reserve1_1 = ((uint256(reserve0) * reserve1) / reserve0_1);
+
+    // Calculate ao0
+    // ao0 = r0 - r0'
+    amountOut0 = reserve0 - reserve0_1;
+
+    // Calculate ai1
+    // ai1 = r1' - r1
+    amountIn1 = reserve1_1 - reserve1;
+  }
+
+  /**
+   * @notice Calculates the price impact on token 0 when the given price impact on token 1 occurs
+   * @param reserve0 The current reserve of token 0
+   * @param reserve1 The current reserve of token 1
+   * @param priceImpact1 The price impact on token 1
+   * @return priceImpact0 The price impact on token 0 when the given price impact on token 1 occurs
+   */
+  function calculateInversePriceImpact(
+    uint128 reserve0,
+    uint128 reserve1,
+    UFixed32x4 priceImpact1
+  ) internal pure returns (UFixed32x4 priceImpact0) {
+    uint256 scalar = 1e33;
+
+    // Calculate p1
+    // p1 = r0 / r1
+    uint256 p1 = (uint256(reserve0) * scalar) / reserve1; // +1e33
+
+    // Calculate p1'
+    // p1' = p1 - (p1 * priceImpact1)
+    uint256 p1_1 = p1 * FixedMathLib.multiplier - p1 * UFixed32x4.unwrap(priceImpact1); // +1e37
+
+    // Calculate pi0
+    // pi0 = (p0' - p0) / p0
+    // pi0 = (1/p1' - 1/p1) / 1/p1
+    // pi0 = (1/p1' - 1/p1) * p1
+    // pi0 = p1/p1' - 1
+    priceImpact0 = UFixed32x4.wrap(
+      uint32(
+        (p1 * FixedMathLib.multiplier * FixedMathLib.multiplier) / p1_1 - FixedMathLib.multiplier
```

Comment:
I think we may be able to remove 1 `FixedMathLib.multiplier` if we use `FixedMathLib.mul` for the calculation of `p1_1` above.